otherlibs/unix/write_unix.c

```diff
@@ -54,6 +55,33 @@ CAMLprim value caml_unix_write(value fd, value buf, value vofs, value vlen)
   CAMLreturn(Val_long(written));
 }
 
+CAMLprim value caml_unix_write_bigarray(value fd, value vbuf, value vofs, value vlen)
+{
+  CAMLparam0();
+  long ofs, len, written;
+  void *buf;
+  int ret;
+
+  buf = Caml_ba_data_val(vbuf);
+  ofs = Long_val(vofs);
+  len = Long_val(vlen);
+  written = 0;
+  caml_enter_blocking_section();
+  while (len > 0) {
+    ret = write(Int_val(fd), buf + ofs, len);
```

Comment:
This is not GC-safe: the bigarray could be collected after `caml_enter_blocking_section`, making `buf` invalid, e.g. in code like this:

```ocaml
Unix.write_bigarray fd (Bigarray.Array1.init ...) pos len
```

It should be sufficient to guard `vbuf` with one of the `CAMLparam*` macros, since the `data` itself won't be moved as long as the underlying bigarray is alive.

The other functions have the same issue (even if it's unlikely to arise in practice for reads).

Reply:
Oops, good catch! It looks like the same bug was also present in the handling of Windows file descriptors (which are custom blocks). I believe both are now fixed. Thanks!

Reply:
Ah, I had missed this comment which was the context for the change in the new commit. I think that rooting `vbuf` is indeed necessary, but you should still access its fields before entering the blocking section.

Reply:
OK. But just for my own understanding: do you agree that the present code (rooting `vbuf` + accesing the fields _inside_ the blocking section) is safe as it is?

Reply:
But I think that accessing arguments of `vbuf` inside the blocking section is *not* correct. My understanding is that you need to have the runtime lock to read inside a memory block owned by the GC, otherwise you may race with a GC moving the value around. (This is consistent with the comment in lseek_unix that is quoted in the other thread.)

Reply:
Right, makes sense. I went back to accessing the blocks outside the runtime block. Hopefully the code is correct now.

Reply:
I believe that the code is correct. This said, I am not sure why you are not following the "full" discipline of registering all the value parameters (even those that we assume are immediate). It would make the code easier to read, and I would bet that there is no noticeable performance differrence.

Reply:
> It would make the code easier to read,

Done. Thanks.

Reply:
@gasche is correct with this description:
> I think that accessing arguments of vbuf inside the blocking section is not correct. My understanding is that you need to have the runtime lock to read inside a memory block owned by the GC, otherwise you may race with a GC moving the value around.

The issue is fixed here, but still present in `read_win32.c` and `write_win32.c` - since `Descr_val(vfd)` points into GC-managed memory, it cannot be used inside a blocking section.

Reply:
Of course! Pushed a fix for the Win32 stubs.